### ISSUE TITLE
scroll events don't bubble, so useCapture makes more sense

### ISSUE THREAD

**Issue by tommie** [PolymerElements/iron-scroll-target-behavior#11]
If an element of the page handles scrolling, like in the scrolling-region demo, the scroll event will not bubble up to the document per specification: https://www.w3.org/TR/DOM-Level-3-Events/#event-type-scroll

This means a user of this behavior that wants to capture all scroll events in a document is out of luck.

I propose setting `useCapture` to true in `addEventListener` for the default case when `scrollTarget` is the document.

I'm not fully comfortable with JS event propagation, so please correct me if I've misunderstood something.

**Comment by blasten** [comment]
> This means a user of this behavior that wants to capture all scroll events in a document is out of luck.

what would the use case be?

**Comment by tommie** [comment]
I want to implement "infinite scrolling" by finding out when my end-of-list element is within the `window` viewport. I don't care about the scrolling per-se, just that _something_ (like scroll or window resize) has happened that may have caused the end-of-list element to become visible. Thus, this is just a proxy event for what I'm really interested in.

The website I'm embedding in already has a scrolling main pane outside of my control, and I'd prefer if my end-of-list element didn't have to know/care about which element in the hierarchy it is.

I could go and do `window.addEventListener('scroll', ..., true)` on my own, of course, but it felt like changing `i-s-t-b` to `useCapture` was a non-invasive change that would make it useful in more use-cases.

**Comment by blasten** [comment]
@tommie have you tried using iron-scroll-threshold? https://github.com/PolymerElements/iron-scroll-threshold

**Comment by tommie** [comment]
I looked at it. It doesn't use the scroll event at all, and triggers a timer every 200 ms, which is not going to be very nice from a battery perspective. I'd like to avoid polling if possible.

I've solved my problem using `window.addEventListener` directly, so feel free to close this issue if you think it's out-of-scope for `<iron-scroll-target-behavior>`.

**Comment by blasten** [comment]
> It doesn't use the scroll event at all,

It does, it's called `_scrollHandler`. The method that this behavior provides.

> and triggers a timer every 200 ms, which is not going to be very nice from a battery perspective. I'd like to avoid polling if possible.

well, the idea was to avoid repetitive work on the scroll event that might force layout and that it's not necessary for correctness. I guess the other approach is to use `requestAnimationFrame` to ensure that layout has happened, but it doesn't avoid the 2nd issue.  Q: how did you infer that the current implementation is not good from a battery perspective?

**Comment by tommie** [comment]
> Q: how did you infer that the current implementation is not good from a battery perspective?

Didn't read the code correctly. I read it like it was using `debounce` to schedule another `debounce` call. My bad.

But since it actually uses `iron-scroll-target`, I don't see how it would help the problem that scroll events don't bubble to the window/document node. It will never receive a `_scrollHandler` call if the scroll target is set to the document but some child element is `overflow-y: scroll`.

**Comment by blasten** [comment]
That's right. I'm still not sure why you want a single scrollHandler that captures all the scroll events. Are you doing that for performance? How many `overflow-y:auto` divs do you have?

**Comment by tommie** [comment]
Not for performance but for code simplicity.

> How many overflow-y:auto divs do you have?

Just a single one. The document structure is something like

``` html
<div>
  <div>
    <div>
      <some-long-list></some-long-list>
      <eol-marker visible="{{visible}}"></eol-marker>
```

I'm implementing `<eol-marker>` to set `visible` to whether it's visible in the window viewport. The problem is that element doesn't know which of the parent `div`s does the scrolling. So rather than having to hard-code that in `<eol-marker>`, I just listen to document events. I don't care about the specific scroll events, just that something may have changed in the document.

Basically, I don't want to have to know the `scrollTarget`, and since `IronScrollTargetBehavior` sets it to the document by default, I had assumed it would Just Work (TM). It doesn't, because the scroll events of arbitrary `div`s don't bubble.

**Comment by tommie** [comment]
Turns out listening to the `document` does not work well with Shadow DOM (but shady DOM is fine), because [the `scroll` event is scoped](https://www.w3.org/TR/shadow-dom/#events-that-are-not-leaked-into-ancestor-trees). To solve this without requiring setting `scrollTarget`, I'd have to register listeners along all shadow roots up to the document node. Doable, but more work than just setting `useCapture`, which was my suggestion here.
